File: app/src/main/java/bassamalim/hidaya/activities/QuranSearcherActivity.java
package bassamalim.hidaya.activities;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.style.ForegroundColorSpan;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Spinner;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.room.Room;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import bassamalim.hidaya.R;
import bassamalim.hidaya.adapters.QuranSearcherAdapter;
import bassamalim.hidaya.database.AppDatabase;
import bassamalim.hidaya.database.dbs.AyatDB;
import bassamalim.hidaya.databinding.ActivityQuranSearcherBinding;
import bassamalim.hidaya.models.Ayah;
import bassamalim.hidaya.other.Utils;

public class QuranSearcherActivity extends AppCompatActivity {

    private ActivityQuranSearcherBinding binding;
    private SharedPreferences pref;
    private RecyclerView recyclerView;
    private SearchView searchView;
    private QuranSearcherAdapter adapter;
    private List<AyatDB> allAyat;
    private List<Ayah> matches;
    private int maxMatches;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Utils.myOnActivityCreated(this);
        binding = ActivityQuranSearcherBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        binding.home.setOnClickListener(v -> onBackPressed());

        init();

        setListeners();

        setupSizeSpinner();

        initRecycler();
    }

    private void init() {
        AppDatabase db = Room.databaseBuilder(this, AppDatabase.class, "HidayaDB")
                .createFromAsset("databases/HidayaDB.db").allowMainThreadQueries().build();
        pref = PreferenceManager.getDefaultSharedPreferences(this);

        searchView = binding.searchView;

        allAyat = db.ayahDao().getAll();

        matches = new ArrayList<>();

        maxMatches = pref.getInt("quran_searcher_matches_last_position", 1);
    }

    private void setListeners() {
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                perform(query);
                return true;
            }
            @Override
            public boolean onQueryTextChange(String newText) {
                return true;
            }
        });
    }

    private void setupSizeSpinner() {
        Spinner spinner = binding.sizeSpinner;

        int last = pref.getInt("quran_searcher_matches_last_position", 0);
        spinner.setSelection(last);

        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long vId) {
                maxMatches = Integer.parseInt(spinner.getItemAtPosition(position).toString());

                SharedPreferences.Editor editor = pref.edit();
                editor.putInt("quran_searcher_matches_last_position", position);
                editor.apply();

                if (adapter != null && adapter.getItemCount() > 0)
                    perform(searchView.getQuery().toString());
            }
            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });
    }

    private void perform(String query) {
        search(query);

        if (matches.isEmpty()) {
            binding.notFoundTv.setVisibility(View.VISIBLE);
            binding.recycler.setVisibility(View.INVISIBLE);
        }
        else {
            binding.notFoundTv.setVisibility(View.INVISIBLE);
            binding.recycler.setVisibility(View.VISIBLE);

            if (adapter != null)
                adapter.notifyDataSetChanged();
        }
    }

    private void search(String text) {
        matches.clear();

        for (int i = 0; i < allAyat.size(); i++) {
            AyatDB a = allAyat.get(i);

            Matcher m = Pattern.compile(text).matcher(a.getAya_text_emlaey());
            SpannableString ss = new SpannableString(a.getAya_text_emlaey());
            while (m.find()) {
                ss.setSpan(new ForegroundColorSpan(getColor(R.color.highlight_M)),
                        m.start(), m.end(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);

                matches.add(new Ayah(a.getSura_no(), a.getSura_name_ar(), a.getPage(),
                        a.getAya_no(), a.getAya_tafseer(), ss));
            }

            if (matches.size() == maxMatches)
                break;
        }
    }

    private void initRecycler() {
        recyclerView = binding.recycler;
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        adapter = new QuranSearcherAdapter(this, matches);
        recyclerView.setAdapter(adapter);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        binding = null;
        if (recyclerView != null)
            recyclerView.setAdapter(null);
        adapter = null;
    }
}

File: app/src/main/java/bassamalim/hidaya/activities/QuranSearcherActivity.kt
package bassamalim.hidaya.activities

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.items
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.preference.PreferenceManager
import bassamalim.hidaya.R
import bassamalim.hidaya.database.dbs.AyatDB
import bassamalim.hidaya.models.QuranSearcherMatch
import bassamalim.hidaya.ui.components.*
import bassamalim.hidaya.ui.theme.AppTheme
import bassamalim.hidaya.utils.ActivityUtils
import bassamalim.hidaya.utils.DBUtils
import bassamalim.hidaya.utils.LangUtils
import java.util.regex.Pattern

class QuranSearcherActivity : ComponentActivity() {

    private lateinit var pref: SharedPreferences
    private lateinit var allAyat: List<AyatDB?>
    private val matches = mutableStateListOf<QuranSearcherMatch>()
    private lateinit var names: List<String>
    private val maxMatchesIndex = mutableStateOf(0)
    private lateinit var maxMatchesItems: Array<String>
    private lateinit var language: String
    private var searched = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        language = ActivityUtils.onActivityCreateSetLocale(this)

        init()

        setContent {
            AppTheme {
                UI()
            }
        }
    }

    private fun init() {
        val db = DBUtils.getDB(this)
        pref = PreferenceManager.getDefaultSharedPreferences(this)

        allAyat = db.ayahDao().getAll()

        names =
            if (language == "en") db.suarDao().getNamesEn()
            else db.suarDao().getNames()

        initMaxMatches()
    }

    private fun initMaxMatches() {
        maxMatchesIndex.value = pref.getInt("quran_searcher_max_matches_index", 0)

        maxMatchesItems =
            if (language == "en") resources.getStringArray(R.array.searcher_matches_en)
            else resources.getStringArray(R.array.searcher_matches)
    }

    private fun search(text: String, highlightColor: Color) {
        matches.clear()

        for (i in allAyat.indices) {
            val a = allAyat[i]!!
            val string = a.aya_text_emlaey

            val matcher = Pattern.compile(text).matcher(string)
            if (matcher.find()) {
                val annotatedString = buildAnnotatedString {
                    append(string)

                    do {
                        addStyle(
                            style = SpanStyle(color = highlightColor),
                            start = matcher.start(),
                            end = matcher.end()
                        )
                    } while (matcher.find())
                }

                matches.add(
                    QuranSearcherMatch(
                        a.sura_num, a.aya_num, names[a.sura_num], a.page,
                        annotatedString, a.aya_tafseer
                    )
                )

                searched = true
                if (matches.size == maxMatchesItems[maxMatchesIndex.value].toInt()) return
            }
        }
    }

    @Composable
    private fun UI() {
        val textState = remember { mutableStateOf(TextFieldValue("")) }
        val highlightColor = AppTheme.colors.accent

        MyScaffold(stringResource(id = R.string.quran_searcher)) { padding ->
            Column(
                Modifier
                    .fillMaxSize()
                    .padding(padding),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Column(
                    Modifier
                        .fillMaxWidth()
                        .height(200.dp),
                    verticalArrangement = Arrangement.SpaceAround,
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    MyText(
                        text = stringResource(id = R.string.search_for_quran_text)
                    )

                    SearchComponent(
                        state = textState,
                        hint = stringResource(id = R.string.search),
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(horizontal = 30.dp),
                        onSubmit = {
                            search(
                                text = textState.value.text,
                                highlightColor = highlightColor
                            )
                        }
                    )

                    Row(
                        Modifier.fillMaxWidth(),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.SpaceEvenly
                    ) {
                        MyText(
                            text = stringResource(id = R.string.max_num_of_marches),
                            fontSize = 18.sp
                        )

                        MyDropDownMenu(
                            selectedIndex = maxMatchesIndex,
                            items = maxMatchesItems
                        ) { index ->
                            pref.edit()
                                .putInt("books_searcher_max_matches_index", index)
                                .apply()
                        }
                    }
                }

                if (matches.isEmpty()) {
                    if (searched)
                        MyText(
                            text = stringResource(id = R.string.no_matches),
                            modifier = Modifier.padding(top = 100.dp)
                        )
                }
                else {
                    MyLazyColumn(lazyList = {
                        items(
                            items = matches
                        ) { item ->
                            MySurface {
                                Column(
                                    Modifier.padding(vertical = 4.dp, horizontal = 4.dp),
                                    horizontalAlignment = Alignment.CenterHorizontally
                                ) {
                                    Row(
                                        Modifier
                                            .fillMaxWidth()
                                            .padding(6.dp),
                                        horizontalArrangement = Arrangement.SpaceEvenly
                                    ) {
                                        MyText(
                                            "${stringResource(R.string.sura)} ${item.suraName}"
                                        )

                                        MyText(
                                            "${stringResource(R.string.page)} " +
                                                    LangUtils.translateNums(
                                                        this@QuranSearcherActivity,
                                                        item.pageNum.toString(),
                                                        false
                                                    )
                                        )
                                    }

                                    MyText(
                                        text = "${stringResource(id = R.string.aya_number)} " +
                                            LangUtils.translateNums(
                                                this@QuranSearcherActivity,
                                                item.ayaNum.toString(),
                                                false
                                            ),
                                        modifier = Modifier.padding(6.dp)
                                    )
                                    MyText(
                                        text = item.text,
                                        modifier = Modifier.padding(6.dp)
                                    )
                                    MyText(
                                        text = "${stringResource(R.string.tafseer)}: ${item.tafseer}",
                                        modifier = Modifier.padding(6.dp)
                                    )

                                    MyButton(
                                        text = stringResource(id = R.string.go_to_page),
                                        textColor = AppTheme.colors.accent,
                                        elevation = 0,
                                        innerPadding = PaddingValues(0.dp),
                                        modifier = Modifier.padding(bottom = 6.dp)
                                    ) {
                                        val intent = Intent(
                                            this@QuranSearcherActivity,
                                            QuranViewer::class.java
                                        )
                                        intent.action = "by_page"
                                        intent.putExtra("page", item.pageNum)
                                        startActivity(intent)
                                    }
                                }
                            }
                        }
                    })
                }
            }
        }
    }
}